src/librbd/crypto/CryptoObjectDispatch.cc

```diff
@@ -29,57 +29,66 @@ struct C_EncryptedObjectReadRequest : public Context {
     I* image_ctx;
     CryptoInterface* crypto;
     uint64_t object_no;
-    uint64_t object_off;
-    ceph::bufferlist* read_data;
+    io::ReadExtents* extents;
     Context* onfinish;
     io::ReadResult::C_ObjectReadRequest* req_comp;
     io::ObjectDispatchSpec* req;
 
     C_EncryptedObjectReadRequest(
             I* image_ctx, CryptoInterface* crypto, uint64_t object_no,
-            uint64_t object_off, uint64_t object_len, IOContext io_context,
+            io::ReadExtents* extents, IOContext io_context,
             int op_flags, const ZTracer::Trace &parent_trace,
-            ceph::bufferlist* read_data, int* object_dispatch_flags,
+            uint64_t* version, int* object_dispatch_flags,
             Context** on_finish,
             Context* on_dispatched) : image_ctx(image_ctx),
                                       crypto(crypto),
                                       object_no(object_no),
-                                      object_off(object_off),
-                                      read_data(read_data),
+                                      extents(extents),
                                       onfinish(on_dispatched) {
       *on_finish = util::create_context_callback<
               Context, &Context::complete>(*on_finish, crypto);
 
-      auto aio_comp = io::AioCompletion::create_and_start(
-              (Context*)this, util::get_image_ctx(image_ctx),
-              io::AIO_TYPE_READ);
-      aio_comp->read_result = io::ReadResult{read_data};
-      aio_comp->set_request_count(1);
-
-      auto req_comp = new io::ReadResult::C_ObjectReadRequest(
-              aio_comp, object_off, object_len, {{0, object_len}});
-
       req = io::ObjectDispatchSpec::create_read(
               image_ctx, io::OBJECT_DISPATCH_LAYER_CRYPTO, object_no,
-              {{object_off, object_len}}, io_context, op_flags, parent_trace,
-              &req_comp->bl, &req_comp->extent_map, nullptr, req_comp);
+              &req_comp->extents, io_context, op_flags, parent_trace,
+              version, this);
     }
 
     void send() {
       req->send();
     }
 
     void finish(int r) override {
-      ldout(image_ctx->cct, 20) << "r=" << r << dendl;
-      if (r > 0) {
-        auto crypto_ret = crypto->decrypt(
-                read_data,
-                Striper::get_file_offset(
-                        image_ctx->cct, &image_ctx->layout, object_no,
-                        object_off));
-        if (crypto_ret != 0) {
-          ceph_assert(crypto_ret < 0);
-          r = crypto_ret;
+      auto cct = image_ctx->cct;
+      ldout(cct, 20) << "r=" << r << dendl;
+      if (r == 0) {
+        for (auto& extent: *extents) {
+          Striper::StripedReadResult destriper;
+          bufferlist bl;
+
+          if (!extent.extent_map.empty()) {
```

Comment:
Nit: it's probably subtle, but why does this need to have its own destriper vs just re-use the logic from `C_ObjectReadRequest`?

Reply:
Do you suggest creating a manual C_ObjectReadRequest + AioCompletion objects per each extent, and then manually calling the finish method?

Reply:
Yeah. Especially since you might need to align the provide extents, you can create a new `C_ObjectReadRequest` with a potentially new set of extents (and use a new `AioCompletion` to write it).

Additionally, I think the `ReadExtents` passed to the read method should have object-extent buffer offsets instead of the image-extent provided buffer offsets since you really aren't trying to destripe at this layer but rather de-sparsify (right?).

Reply:
I'm now re-thinking this and I don't think it's a good idea.
This will be longer code with wasted copy ops.
How about if I refactor the common code (calling destriper.add_partial...) to one place (say, in librbd/io/Utils)?

Reply:
That would possibly help, but I think I still don't understand how this will be re-assembled correctly with the current logic:

1) IO comes in via image dispatcher (i.e. read 0~2048)
2) Assuming a stripe count of 2 and a stripe unit of 512, this would result in two object requests:
  a) object 0: read 0~512,512~512 into buffers 0~512,1024~512
  b) object 1: read 0~512,512~512 into buffers 512~512, 1536~512
3) This layer takes those read extents and buffer offsets and tries to build a bl based on the image-extent bufferlist
  a) object 0: creates a bl from 0~1536 (zeroed from 512~1024)
  b) object 1: creates a bl from 0~2048 (zeroed from 0~512,1024~512)

I think for you to re-use the destriper, you need to build your own `ReadExtents` in `io::ObjectDispatchSpec::create_read` so that the buffer offsets remain in object-extent space.

Reply:
First, looking at ImageReadRequest<I>::send_request, each object request is per a single extent. I do not see any batching per object.

Second, even if the crypto layer got a read request with multiple extents, it creates a bl per extent, not per object request, as you described in step 3 above.

Reply:
My example was a single image extent `0~2048` but with striping enabled. The `ReadExtents` that are passed down for object 0 in this example would include the buffer extents that would re-order the data into the image-extent bufferlist, not keep it in object-extents for decryption. All I am saying is I think your buffer_extents are incorrect at this point.

Reply:
Finally got you!!
Simple fix on the way... (hopefully :))

Reply:
ready again for review